client/client.go

```diff
@@ -99,56 +123,513 @@ func UploadAdd(issuer principal.Signer, space did.DID, params uploadadd.Caveat,
 	}
 
 	resp, err := client.Execute([]invocation.Invocation{inv}, cfg.conn)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("sending invocation: %w", err)
 	}
 
 	rcptlnk, ok := resp.Get(inv.Link())
 	if !ok {
-		return nil, fmt.Errorf("receipt not found: %s", inv.Link())
+		return nil, nil, fmt.Errorf("receipt not found: %s", inv.Link())
 	}
 
-	reader, err := uploadlist.NewReceiptReader()
+	// reader, err := receipt.NewReceiptReaderFromTypes[spaceblob.AddOk, fdm.FailureModel](spaceblob.AddOkType(), fdm.FailureType())
+	// if err != nil {
+	// 	return nil, nil, fmt.Errorf("generating receipt reader: %w", err)
+	// }
+	reader := receipt.NewAnyReceiptReader(captypes.Converters...)
+
+	rcpt, err := reader.Read(rcptlnk, resp.Blocks())
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("reading receipt: %w", err)
 	}
 
-	return reader.Read(rcptlnk, resp.Blocks())
+	//_, err = result.Unwrap(result.MapError(rcpt.Out(), failure.FromFailureModel))
+	_, fail := result.Unwrap(rcpt.Out())
+	if fail != nil {
+		return nil, nil, fmt.Errorf("blob add failed: %v", fail)
+	}
+
+	var allocateTask, putTask, acceptTask invocation.Invocation
+	legacyAccept := false
+	var concludeFxs []invocation.Invocation
+	for _, task := range rcpt.Fx().Fork() {
+		inv, ok := task.Invocation()
+		if ok {
+			switch inv.Capabilities()[0].Can() {
+			case blob.AllocateAbility:
+				allocateTask = inv
+			case w3sBlob.AllocateAbility:
+				if allocateTask == nil {
+					allocateTask = inv
+				}
+			case ucancap.ConcludeAbility:
+				concludeFxs = append(concludeFxs, inv)
+			case httpcap.PutAbility:
+				putTask = inv
+			case blob.AcceptAbility:
+				acceptTask = inv
+			case w3sBlob.AcceptAbility:
+				if acceptTask == nil {
+					acceptTask = inv
+					legacyAccept = true
+				}
+			}
+		}
+	}
+
+	if allocateTask == nil || len(concludeFxs) == 0 || putTask == nil || acceptTask == nil {
+		return nil, nil, fmt.Errorf("mandatory tasks not received in space/blob/add receipt")
+	}
+
+	var allocateRcpt receipt.Receipt[blob.AllocateOk, fdm.FailureModel]
+	var legacyAllocateRcpt receipt.Receipt[w3sBlob.AllocateOk, fdm.FailureModel]
+	var putRcpt receipt.AnyReceipt
+	var acceptRcpt receipt.Receipt[blob.AcceptOk, fdm.FailureModel]
+	var legacyAcceptRcpt receipt.Receipt[w3sBlob.AcceptOk, fdm.FailureModel]
+	for _, concludeFx := range concludeFxs {
+		concludeRcpt, err := getConcludeReceipt(concludeFx)
+		if err != nil {
+			return nil, nil, fmt.Errorf("reading ucan/conclude receipt: %w", err)
+		}
+
+		switch concludeRcpt.Ran().Link() {
+		case allocateTask.Link():
+			switch allocateTask.Capabilities()[0].Can() {
+			case blob.AllocateAbility:
+				allocateRcpt, err = receipt.Rebind[blob.AllocateOk, fdm.FailureModel](concludeRcpt, blob.AllocateOkType(), fdm.FailureType(), captypes.Converters...)
+				if err != nil {
+					return nil, nil, fmt.Errorf("bad allocate receipt in conclude fx: %w", err)
+				}
+			case w3sBlob.AllocateAbility:
```

Comment:
There is a LOT of ceremony/complexity we're doing around distinguishing between legacy allocate vs allocate, same with accept / legacy accept.

These are actually identical structs. I wonder if the solution is to just modify go-libstoracha so that the receipts for legacy just uses the type from regular blob capabilities imported

Reply:
alternatively, @alanshaw do we even need to support the web3.storage capabilities in the go-client? Is this for any other reason that the upload service is not in prod?

Reply:
we still need to support `web3.storage/blob` because this is what legacy spaces use. But yeah, I was thinking of creating some common interface in `go-libstoracha/capabilities` to allow handling them indistinctly, but merging them makes sense to me and it's a more straightforward approach.

I captured this in https://github.com/storacha/go-libstoracha/issues/19.

Reply:
No because an old space will not go through warm storage nodes...